### PR TITLE
Rewrite Connection Manager to Handle Multiple Connections

### PR DESCRIPTION
**Pull request by swood-scottlogic**
## Description
This PR just handles the changes made to the existing web socket code. The previous implementation made it very difficult to send data on the web socket unless it was in response to a message received. The new implementation starts to handle the idea of having multiple connections active at one time, but will require more work to be done on the implementation of user sessions.

## Changelog
- Reworks connection manager to handle multiple web sockets
- Adds new broadcast method to send Messages from the backend on the web socket
- Removes on_message method and recreates functionality in app.py
    - Functionality changed from using recursion to a while loop as long running connections could cause a recursion depth exceeded exception
- Rewrites unit tests to better cover new structure

### CODE REVIEW

**Review by gaganahluwalia — DISMISSED** [review]
Looks all good to me.

File: backend/tests/api/connection_manager_test.py [hunk 1]
```diff
@@ -1,84 +1,159 @@
...
+    mock_ws.accept = mocker.AsyncMock()
+    mock_ws.send_json = mocker.AsyncMock()
+    mock_on_message = mocker.AsyncMock
+
+    return manager, mock_ws, mock_on_message
+
+
+@pytest.mark.asyncio
+async def test_connect_new_websocket(connection_manager):
+    manager, mock_ws, _ = connection_manager
```
Comment (hsauve-scottlogic) [inline backend/tests/api/connection_manager_test.py:29]: what does the underscore represent here please?

**Review by hsauve-scottlogic — APPROVED** [review]
Looks good to me, and all tests pass on my machine, I just left one comment for clarification, thanks

File: backend/tests/api/connection_manager_test.py (hunk 1 above)
Comment (swood-scottlogic) [reply backend/tests/api/connection_manager_test.py:29]: The `connection_manager` function defined on line 17 returns three things: `manager, mock_ws, mock_on_message`. In this test I don't need the `mock_on_message` value that has been returned and one convention to signify I won't be using this is to assign it to the variable `_`. I've had a look and can't find an official source about this convention but it is referenced in various places. The additional benefit is pylance (the language server vscode uses for warnings and errors in python)  won't report the `_` variable as being unused.

**Review by gaganahluwalia — APPROVED** [review]

**Review by hsauve-scottlogic — APPROVED** [review]

**Pull request merged**